pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java

```diff
@@ -66,6 +68,10 @@ protected ConnectionHandler(HandlerState state, Backoff backoff, Connection conn
     }
 
     protected void grabCnx() {
+        grabCnx(Optional.empty());
+    }
+
+    protected void grabCnx(Optional<URI> hostURI) {
```

Comment:
Maybe we can add `@Nonnull` annotation for param `Optional<URI> hostUrl`, or else we still need to check if it's null.